halo2/src/prover.rs

```diff
@@ -74,13 +96,16 @@ impl Halo2Prover {
 
         log::info!("Starting proof generation...");
 
-        let (circuit, publics) = analyzed_to_circuit(pil, fixed, witness);
+        let (plaf_circuit, circuit_data) = analyzed_to_circuit(pil, fixed);
+        let (circuit, publics) =
+            analyzed_to_circuit_with_witness(pil, plaf_circuit.plaf, circuit_data, witness);
 
         log::debug!("{}", PlafDisplayBaseTOML(&circuit.plaf));
 
         log::info!("Generating VK and PK for snark...");
-        let vk = keygen_vk(&params, &circuit).unwrap();
-        let pk = keygen_pk(&params, vk, &circuit).unwrap();
+        //let vk = keygen_vk(&params, &circuit).unwrap();
+        let pk = keygen_pk(&params, self.vkey.clone().unwrap(), &circuit).unwrap();
```

Comment:
This is one of the main hacks that I have no idea how to solve.
- we want to pass the verification key to the prover integration so that it can use a previously generated one and not need to compute on the fly
- Halo2 wants the `vk` by value in this function
- so we need to either `clone` or `take` it, not sure which is better

Reply:
I think cloning should be fine? It's supposed to be small, right?

Reply:
that's true, I was thinking in the order of magnitude of the setup